src/postgres_deparse.c

```diff
@@ -130,173 +94,583 @@ isOp(const char *val)
+		}
+
+		state->emitted_comments = bms_add_member(state->emitted_comments, i);
+	}
+}
+
+static void
+deparseEmit(DeparseState *state, StringInfo str)
+{
+	ListCell *lc;
+	foreach (lc, state->result)
+	{
+		DeparseStateSubPart *subpart = (DeparseStateSubPart *) lfirst(lc);
+		bool first_subpart = list_cell_number(state->result, lc) == 0;
+		bool last_subpart = list_cell_number(state->result, lc) == list_length(state->result) - 1;
+
+		if (state->opts.pretty_print && !first_subpart && (!last_subpart || subpart->str->len > 0))
+		{
+			appendStringInfoChar(str, '\n');
+		}
+		else if (!first_subpart && (state->opts.pretty_print || str->data[str->len - 1] != '('))
+		{
+			appendStringInfoChar(str, ' ');
+		}
+
+		if (state->opts.pretty_print)
+		{
+			for (int i = 0; i < subpart->indent; i++)
+				appendStringInfoChar(str, ' ');
+		}
+		appendStringInfoString(str, subpart->str->data);
+		removeTrailingSpaceFromStr(str);
+
+		pfree(subpart->str->data);
+		pfree(subpart->str);
+		pfree(subpart);
+	}
+	list_free(state->result);
+	state->result = NIL;
+	if (state->opts.pretty_print && state->opts.trailing_newline)
+		appendStringInfoChar(str, '\n');
+}
+
+static void
+deparseStatePushStatement(DeparseState *state)
+{
+	DeparseStateStatement *stmt = palloc0(sizeof(DeparseStateStatement));
+	stmt->parent = state->current;
+	if (stmt->parent)
+	{
+		DeparseStatePart *part = deparseGetCurrentPart(state);
+		stmt->base_indent = stmt->parent->base_indent + state->opts.indent_size;
+		if (part->indent_subparts) /* Indent again if subparts next to us are also indented */
+			stmt->base_indent += state->opts.indent_size;
+	}
+	state->current = stmt;
+}
+
+static void
+deparseStatePopStatement(DeparseState *state)
+{
+	ListCell *lc;
+	ListCell *lc2;
+	DeparseStateStatement *stmt = state->current;
+	Assert(stmt != NULL);
+	List **result = NULL;
+
+	if (stmt->parent)
+	{
+		state->current = stmt->parent;
+		result = &deparseGetCurrentPart(state)->subparts;
+	}
+	else
+	{
+		// If its the top-level statement, save as results instead
+		result = &state->result;
+	}
+
+	// Step through all lines, as if we did the final emit, but instead emit line-by-line into the parent
+	foreach (lc, stmt->parts)
+	{
+		DeparseStatePart *part = (DeparseStatePart *) lfirst(lc);
+		bool first_part = list_cell_number(stmt->parts, lc) == 0;
+
+		/* Only merge subparts if they all have the same indentation level */
+		/* TODO: We can be a bit more permissive here - basically we want to avoid merging across statement levels */
+		bool merge_subparts = part->merge_subparts;
+		int first_indent = ((DeparseStateSubPart *) linitial(part->subparts))->indent;
+		for_each_from (lc2, part->subparts, 1)
+		{
+			DeparseStateSubPart *subpart = (DeparseStateSubPart *) lfirst(lc2);
+			if (subpart->indent != first_indent)
+			{
+				merge_subparts = false;
+				break;
+			}
+		}
+
+		/* Merge subparts */
+		if (state->opts.pretty_print && merge_subparts && list_length(part->subparts) > 1)
```

Comment:
Should we drop the pretty print flag here? (it would help keep the number of parts down for complex statements, even when not pretty printing)